### ISSUE TITLE
Lag in screen mirroring 

### ISSUE THREAD

**Issue by Mahdi-RH** [pedroSG94/RootEncoder#1450]
Hi dear, 
I am using RtspServerDisplay class for mirroring the surface of my Android device. when I start mirroring it, there is a lag (usually 3 seconds or more) showing my device's surface on Tv. I use Exoplayer for playing the stream. is there any way to fix it?

**Comment by Mahdi-RH** [comment]
Yes, I mean delay. Thank you for helping me. Is there any way to reduce the delay of the Media Projection class?

**Comment by pedroSG94** [comment]
Hello,

No, but the delay is less than 1 second. I tested using this command instead of Exoplayer:
```
ffplay -fflags nobuffer rtsp://192.168.0.158:1935/
```
With this command the delay is around 1s. I started a chronometer at same time in my PC and the phone and this is the delay result:
![imagen](https://github.com/pedroSG94/RootEncoder/assets/17556876/18d5c663-b043-4c0d-846f-ef2dff7d8f9f)
As you can see the delay is around 740ms

**Comment by Mahdi-RH** [comment]
Thank you. I reduced the delay in Exoplayer by the link you gave me. At first, it works well, but after a few minutes, the time of delay goes up. Do you have any idea about it?  Do you think Exoplayer is a good tool for mirroring  RTSP streams?

**Comment by pedroSG94** [comment]
Hello,

> Do you have any idea about it?

I'm not sure about it, You can try use UDP instead of TCP to reduce the delay and check if this solve the problem because UDP just drop packets instead of add delay if you have any problem sending it. You should do it in player side because if you use UDP or TCP is player decision. You can try both cases (TCP and UDP):
https://github.com/google/ExoPlayer/issues/9151#issuecomment-874159111
Anyway, I recommend you ask in Exoplayer project about it and they can guide you more.

> Do you think Exoplayer is a good tool for screen mirroring ?

No, exoplayer no aim to be a low delay player. You can check that one of the contributor said that:
https://github.com/google/ExoPlayer/issues/10694#issuecomment-1283809498
But I don't know a RTSP player designed for low delay. Usually all players add buffering.
